tools/integration_tests/operations/parallel_dirops_test.go

```diff
@@ -0,0 +1,447 @@
+	}()
+	go func() {
+		deleteErr = deleteFileFunc(&wg, filePath)
+	}()
+	wg.Wait()
+
+	// Assert either file is looked up first or deleted first
+	assert.ErrorIs(t, deleteErr, nil)
+	_, err := os.Stat(filePath)
+	assert.True(t, os.IsNotExist(err))
+	if lookUpErr == nil {
+		assert.Equal(t, int64(5), fileInfo.Size())
+		assert.Contains(t, fileInfo.Name(), "file1.txt")
+		assert.False(t, fileInfo.IsDir())
+	} else {
+		assert.True(t, os.IsNotExist(lookUpErr))
+	}
+}
+
+func TestParallelLookUpAndRenameSameFile(t *testing.T) {
+	testDir := setup.SetupTestDirectory(DirForOperationTests)
+	// Create directory structure for testing.
+	createDirectoryStructureForParallelDiropsTest(t)
+	lookUpFunc := func(wg *sync.WaitGroup, filePath string) (os.FileInfo, error) {
+		defer wg.Done()
+		fileInfo, err := os.Stat(filePath)
+		return fileInfo, err
+	}
+	renameFunc := func(wg *sync.WaitGroup, oldFilePath string, newFilePath string) error {
+		defer wg.Done()
+		err := os.Rename(oldFilePath, newFilePath)
+		return err
+	}
+	var fileInfo os.FileInfo
+	var lookUpErr, renameErr error
+
+	// Parallel lookup and rename of a file.
+	filePath := path.Join(testDir, "explicitDir1", "file1.txt")
+	newFilePath := path.Join(testDir, "newFile.txt")
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+	go func() {
+		fileInfo, lookUpErr = lookUpFunc(&wg, filePath)
+	}()
+	go func() {
+		renameErr = renameFunc(&wg, filePath, newFilePath)
+	}()
+	wg.Wait()
+
+	// Assert either file is renamed first or looked up first
+	assert.ErrorIs(t, renameErr, nil)
+	newFileInfo, err := os.Stat(newFilePath)
+	assert.ErrorIs(t, err, nil)
+	assert.Contains(t, newFileInfo.Name(), "newFile.txt")
+	assert.False(t, newFileInfo.IsDir())
+	assert.Equal(t, int64(5), newFileInfo.Size())
+	if lookUpErr == nil {
+		assert.Equal(t, int64(5), fileInfo.Size())
+		assert.Contains(t, fileInfo.Name(), "file1.txt")
+		assert.False(t, fileInfo.IsDir())
+	} else {
+		assert.True(t, os.IsNotExist(lookUpErr))
+	}
+}
+
+func TestParallelLookUpAndMkdirSameDir(t *testing.T) {
+	testDir := setup.SetupTestDirectory(DirForOperationTests)
+	// Create directory structure for testing.
+	createDirectoryStructureForParallelDiropsTest(t)
+	lookUpFunc := func(wg *sync.WaitGroup, dirPath string) (os.FileInfo, error) {
+		defer wg.Done()
+		fileInfo, err := os.Stat(dirPath)
+		return fileInfo, err
+	}
+	mkdirFunc := func(wg *sync.WaitGroup, dirPath string) error {
+		defer wg.Done()
+		err := os.Mkdir(dirPath, 0600)
+		return err
+	}
+	var statInfo os.FileInfo
+	var lookUpErr, mkdirErr error
+
+	// Parallel lookup and mkdir of a new directory.
+	dirPath := path.Join(testDir, "newDir")
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+	go func() {
+		statInfo, lookUpErr = lookUpFunc(&wg, dirPath)
+	}()
+	go func() {
+		mkdirErr = mkdirFunc(&wg, dirPath)
+	}()
+	wg.Wait()
+
+	// Assert either directory is created first or looked up first
+	assert.ErrorIs(t, mkdirErr, nil)
+	dirStatInfo, err := os.Stat(dirPath)
```

Comment:
Shall we do this checks in the else block only when the previous lookup failed? Also assert on name of the file

Reply:
Changed